tests/editing/uat-testing/uat_testing.py

```diff
@@ -0,0 +1,509 @@
+            value = row[column]
+            # It checks if the value in a cell is a float and has decimal places.
+            if isinstance(value, float) and value % 1 != 0:
+                # If the value has decimal places, it retrieves the corresponding expected value from the expected output DataFrame.
+                expected_value = df_expected_output.loc[index, column]
+                # If the expected value is also a float and has decimal places, it determines the number of decimal places.
+                if isinstance(expected_value, float) and expected_value % 1 != 0:
+                    expected_decimal_places = len(str(expected_value).split(".")[1])
+                    # The value in the processed output DataFrame is then rounded to the same number of decimal places as the expected value.
+                    # This ensures that the values match when comparing the processed output against the expected output.
+                    df_processed_output.loc[index, column] = round(
+                        value, expected_decimal_places
+                    )
+
+    return df_processed_output
+
+
+# The output_failures function is responsible for printing the details of failures that
+# occurred during the comparison of expected output and processed output. It takes a list of failures
+# as input and iterates over each failure. For each failure, it prints the filename with the error,
+# the reference (if available), the row and column where the value mismatch occurred, the expected output value,
+# and the processed output value.
+def output_failures(failures):
+    # Print all the failures
+    for failure in failures:
+        print("\n")
+        print(
+            "===================================================================================================================="
+        )
+        print(f"Filename with the error '{failure['file_name']}'.")
+        for failure_details in failure["failures"]:
+            if "reference" in failure_details:
+                print(
+                    f"Value mismatch at reference {failure_details['reference']}, row {failure_details['row']} and column {failure_details['col']}."
+                )
+            elif "RU" in failure_details:
+                print(
+                    f"Value mismatch at RU {failure_details['RU']}, row {failure_details['row']} and column {failure_details['col']}."
+                )
+            print(f"Expected output: {failure_details['expected_output']}")
+            print(f"Processed output: {failure_details['processed_output']}")
+            print("\n")
+
+
+def compare_dataframes(df_processed_output, df_expected_output, method, file1):
+    # Comparing nan values is problematic when using the datatest library
+    # so we make all the nan values in the processed and expected output 0
+    # so we can easily compare it without being thrown an error when comparing nan values
+    df_processed_output = df_processed_output.fillna(0)
```

Comment:
I wonder if this potentially creates a problem if the expected value is 0 but the processed value was returned as nan? I think this code would mask that error?  Maybe we could set nan values to -1?  Not necessarily something that needs fixing under this ticket but I'd be interested in your thoughts.

Reply:
Great spot Steve as the initial options I was going to go with were 0 or -1 or any other number that we know wouldn't appear as the processed value and then I settled on 0 for some reason.

I think -1 will be suitable in this occasion. You're right. If the expected value is 0 and the processed value is NaN, using fillna(0) would mask this discrepancy, as both values would become 0 after the operation and the test would false pass.

Setting NaN values to -1 (or any other value that we know won't appear in our data) will be the way to handle this, as it would allow us to distinguish between actual 0 values and values that were originally NaN.

This way, if a value was originally NaN in the processed data but 0 in the expected data (or vice versa), the discrepancy will be obvious when we compare the two DataFrames.